File: scripts/match.swift
import AppKit
import Foundation

func loadGray(_ path: String) -> (pixels: [Float], width: Int, height: Int)? {
    guard let image = NSImage(contentsOfFile: path),
          let tiff = image.tiffRepresentation,
          let rep = NSBitmapImageRep(data: tiff),
          let cg = rep.cgImage else { return nil }
    let w = cg.width
    let h = cg.height
    var raw = [UInt8](repeating: 0, count: w * h * 4)
    guard let ctx = CGContext(
        data: &raw,
        width: w,
        height: h,
        bitsPerComponent: 8,
        bytesPerRow: w * 4,
        space: CGColorSpaceCreateDeviceRGB(),
        bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
    ) else { return nil }
    ctx.draw(cg, in: CGRect(x: 0, y: 0, width: w, height: h))
    var gray = [Float](repeating: 0, count: w * h)
    for i in 0..<(w * h) {
        let o = i * 4
        gray[i] = Float(raw[o]) * 0.299 + Float(raw[o + 1]) * 0.587 + Float(raw[o + 2]) * 0.114
    }
    return (gray, w, h)
}

func resizeNearest(_ src: [Float], sw: Int, sh: Int, dw: Int, dh: Int) -> [Float] {
    var out = [Float](repeating: 0, count: dw * dh)
    for y in 0..<dh {
        let sy = min(sh - 1, y * sh / dh)
        for x in 0..<dw {
            let sx = min(sw - 1, x * sw / dw)
            out[y * dw + x] = src[sy * sw + sx]
        }
    }
    return out
}

func rotate90CW(_ src: [Float], w: Int, h: Int) -> ([Float], Int, Int) {
    let nw = h
    let nh = w
    var out = [Float](repeating: 0, count: nw * nh)
    for y in 0..<h {
        for x in 0..<w {
            out[x * nw + (h - 1 - y)] = src[y * w + x]
        }
    }
    return (out, nw, nh)
}

func prepareNeedle(_ needle: [Float]) -> (idxs: [Int], vals: [Float], mean: Float, std: Float)? {
    var idxs = [Int]()
    var vals = [Float]()
    for i in 0..<needle.count {
        if needle[i] >= 70 {
            idxs.append(i)
            vals.append(needle[i])
        }
    }
    if idxs.count < 14 { return nil }
    let n = Float(idxs.count)
    let mean = vals.reduce(0, +) / n
    let varSum = vals.reduce(0) { $0 + ($1 - mean) * ($1 - mean) } / n
    let std = sqrt(max(Float(1e-2), varSum))
    return (idxs, vals, mean, std)
}

func ncc(hay: [Float], hw: Int, nw: Int, prep: (idxs: [Int], vals: [Float], mean: Float, std: Float), x: Int, y: Int) -> Float {
    let n = Float(prep.idxs.count)
    var pSum: Float = 0
    var pSum2: Float = 0
    for ti in prep.idxs {
        let tx = ti % nw
        let ty = ti / nw
        let pv = hay[(y + ty) * hw + (x + tx)]
        pSum += pv
        pSum2 += pv * pv
    }
    let pMean = pSum / n
    let pVar = max(Float(1e-2), pSum2 / n - pMean * pMean)
    var cov: Float = 0
    for k in 0..<prep.idxs.count {
        let ti = prep.idxs[k]
        let tx = ti % nw
        let ty = ti / nw
        let pv = hay[(y + ty) * hw + (x + tx)]
        cov += (pv - pMean) * (prep.vals[k] - prep.mean)
    }
    cov /= n
    let score = cov / (sqrt(pVar) * prep.std)
    return max(-1, min(1, score))
}

func matchInROI(hay: [Float], hw: Int, hh: Int, needle: [Float], nw: Int, nh: Int,
                x0: Int, y0: Int, x1: Int, y1: Int, step: Int) -> (Float, Int, Int)? {
    guard let prep = prepareNeedle(needle) else { return nil }
    if nw >= hw || nh >= hh { return nil }
    let xs = max(0, x0)
    let ys = max(0, y0)
    let xe = min(hw - nw, x1)
    let ye = min(hh - nh, y1)
    if xs > xe || ys > ye { return nil }
    var best: Float = -2
    var bx = xs
    var by = ys
    var y = ys
    while y <= ye {
        var x = xs
        while x <= xe {
            let s = ncc(hay: hay, hw: hw, nw: nw, prep: prep, x: x, y: y)
            if s > best {
                best = s
                bx = x
                by = y
            }
            x += step
        }
        y += step
    }
    let rx0 = max(xs, bx - step)
    let ry0 = max(ys, by - step)
    let rx1 = min(xe, bx + step)
    let ry1 = min(ye, by + step)
    for yy in ry0...ry1 {
        for xx in rx0...rx1 {
            let s = ncc(hay: hay, hw: hw, nw: nw, prep: prep, x: xx, y: yy)
            if s > best {
                best = s
                bx = xx
                by = yy
            }
        }
    }
    return (best, bx, by)
}

var args = Array(CommandLine.arguments.dropFirst())
var roi: (Double, Double, Double, Double)? = nil
if let i = args.firstIndex(of: "--roi"), i + 4 < args.count {
    roi = (Double(args[i + 1]) ?? 0, Double(args[i + 2]) ?? 0, Double(args[i + 3]) ?? 1, Double(args[i + 4]) ?? 1)
    args.removeSubrange(i...(i + 4))
}
guard args.count >= 2 else {
    fputs("usage: match <hay.png> <needle.png> [minScore] [--roi x y w h]\n", stderr)
    exit(1)
}
let hayPath = args[0]
let needlePath = args[1]
let minScore = args.count >= 3 ? (Float(args[2]) ?? 0.75) : 0.75

guard let hay = loadGray(hayPath), let needle0 = loadGray(needlePath) else {
    fputs("ERR:cannot-read\n", stderr)
    exit(2)
}

var templates: [([Float], Int, Int)] = [(needle0.pixels, needle0.width, needle0.height)]
var cur = (needle0.pixels, needle0.width, needle0.height)
for _ in 1...3 {
    let r = rotate90CW(cur.0, w: cur.1, h: cur.2)
    templates.append(r)
    cur = r
}

let rx = roi?.0 ?? 0
let ry = roi?.1 ?? 0
let rw = roi?.2 ?? 1
let rh = roi?.3 ?? 1
let x0 = Int(Double(hay.width) * rx)
let y0 = Int(Double(hay.height) * ry)
let x1 = Int(Double(hay.width) * (rx + rw))
let y1 = Int(Double(hay.height) * (ry + rh))

var best: (Float, Double, Double, Double)? = nil
let scales: [Double] = [0.85, 1.0, 1.2, 1.45, 1.7, 2.0, 2.4, 2.9]
for tpl in templates {
    for scale in scales {
        let tw = max(12, Int(Double(tpl.1) * scale))
        let th = max(12, Int(Double(tpl.2) * scale))
        if tw >= hay.width || th >= hay.height { continue }
        let scaled = resizeNearest(tpl.0, sw: tpl.1, sh: tpl.2, dw: tw, dh: th)
        let step = max(1, min(3, max(tw, th) / 10))
        if let hit = matchInROI(hay: hay.pixels, hw: hay.width, hh: hay.height, needle: scaled, nw: tw, nh: th, x0: x0, y0: y0, x1: x1 - tw, y1: y1 - th, step: step) {
            let cx = Double(hit.1) + Double(tw) / 2.0
            let cy = Double(hit.2) + Double(th) / 2.0
            if best == nil || hit.0 > best!.0 {
                best = (hit.0, cx, cy, scale)
            }
        }
    }
}

guard let best = best, best.0 >= minScore else {
    if let best = best {
        print(String(format: "none\t%.3f", best.0))
    } else {
        print("none")
    }
    exit(0)
}

let fx = best.1 / Double(hay.width)
let fy = best.2 / Double(hay.height)
print(String(format: "%.4f\t%.4f\t%.3f\t%.2f", fx, fy, best.0, best.3))

File: scripts/match.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""在截图里找返回图标。支持多模板、横竖屏 ROI。

输出: fx fy score scale template
fx/fy 相对**截图像素归一化**（screencapture -o 去阴影后）。
调用方必须经 coords.shot_norm_to_screen → screen_to_outer 再点。
ROI 仍按截图比例，与 outer 对齐时即 outer ROI。
"""

import math
import os
import sys

from PIL import Image


def prepare(im, thr=75):
    w, h = im.size
    px = im.load()
    idxs = []
    vals = []
    for y in range(h):
        for x in range(w):
            v = px[x, y]
            if v >= thr:
                idxs.append((x, y))
                vals.append(float(v))
    # 黑底浅色图标：亮点太少时降低阈值再试
    if len(idxs) < 14 and thr > 35:
        return prepare(im, thr=35)
    if len(idxs) < 10:
        return None
    n = float(len(vals))
    mean = sum(vals) / n
    std = math.sqrt(max(1e-2, sum((v - mean) ** 2 for v in vals) / n))
    return idxs, vals, mean, std, w, h


def ncc_at(hay_px, prep, ox, oy):
    idxs, vals, tmean, tstd, _, _ = prep
    n = float(len(idxs))
    samples = [float(hay_px[ox + x, oy + y]) for x, y in idxs]
    pmean = sum(samples) / n
    pvar = sum((v - pmean) ** 2 for v in samples) / n
    cov = sum((samples[i] - pmean) * (vals[i] - tmean) for i in range(len(vals))) / n
    return max(-1.0, min(1.0, cov / (math.sqrt(max(1e-2, pvar)) * tstd)))


def search(hay, prep, roi, step=2):
    hw, hh = hay.size
    hay_px = hay.load()
    _, _, _, _, nw, nh = prep
    x0 = int(hw * roi[0])
    y0 = int(hh * roi[1])
    x1 = min(hw - nw, int(hw * (roi[0] + roi[2])) - nw)
    y1 = min(hh - nh, int(hh * (roi[1] + roi[3])) - nh)
    if x0 > x1 or y0 > y1:
        return None
    cands = []
    y = y0
    while y <= y1:
        x = x0
        while x <= x1:
            s = ncc_at(hay_px, prep, x, y)
            if s >= 0.55:
                cands.append((s, x, y))
            x += step
        y += step
    if not cands:
        return None
    cands.sort(reverse=True)
    best = None
    for s0, bx, by in cands[:8]:
        local = (s0, bx, by)
        for yy in range(max(y0, by - step), min(y1, by + step) + 1):
            for xx in range(max(x0, bx - step), min(x1, bx + step) + 1):
                s = ncc_at(hay_px, prep, xx, yy)
                if s > local[0]:
                    local = (s, xx, yy)
        if best is None or local[0] > best[0]:
            best = local
            if best[0] >= 0.97:
                return best
    return best


def land_rois():
    # 横屏：公告关闭多在右上；偶发左上/右下（镜像旋转）
    return [
        (0.82, 0.00, 0.18, 0.22),
        (0.72, 0.00, 0.28, 0.30),
        (0.00, 0.00, 0.22, 0.22),
        (0.78, 0.70, 0.22, 0.30),
    ]


def port_rois():
    # 竖屏窗：游戏横屏时关闭键常落在底右/底左；竖屏公告则在顶右
    return [
        (0.55, 0.70, 0.42, 0.28),
        (0.40, 0.72, 0.55, 0.26),
        (0.72, 0.00, 0.28, 0.22),
        (0.00, 0.70, 0.35, 0.28),
        (0.00, 0.00, 0.28, 0.22),
    ]


def announce_rois(portrait):
    # 少 ROI、先扫常见角，加快速度（避免点完又卡好几秒）
    if portrait:
        return [
            (0.70, 0.00, 0.28, 0.22),
            (0.55, 0.72, 0.42, 0.26),
            (0.00, 0.72, 0.35, 0.26),
            (0.00, 0.00, 0.28, 0.22),
        ]
    return [
        (0.78, 0.00, 0.22, 0.22),
        (0.00, 0.00, 0.22, 0.22),
        (0.78, 0.75, 0.22, 0.25),
        (0.00, 0.75, 0.22, 0.25),
    ]


def continue_rois(portrait):
    # 「继续游戏」：竖屏偏中下；横屏/竖排字偏右侧
    if portrait:
        return [
            (0.20, 0.55, 0.60, 0.40),
            (0.55, 0.35, 0.40, 0.50),
            (0.10, 0.40, 0.40, 0.50),
        ]
    return [
        (0.55, 0.20, 0.42, 0.60),
        (0.20, 0.55, 0.60, 0.40),
        (0.05, 0.20, 0.40, 0.60),
    ]


def _scaled_jobs(im, path, rois, scales, tag=""):
    out = []
    for sc in scales:
        tw = max(8, int(im.size[0] * sc))
        th = max(8, int(im.size[1] * sc))
        name = path if not tag else (path + tag)
        out.append((im.resize((tw, th), Image.BILINEAR), sc, rois, name))
    return out


def collect_templates(paths, portrait):
    """按方向选模板；公告关闭会四向旋转。"""
    jobs = []
    seen = set()
    for path in paths:
        if not path or not os.path.isfile(path):
            continue
        key = os.path.abspath(path)
        if key in seen:
            continue
        seen.add(key)
        base = os.path.basename(path).lower()
        try:
            im = Image.open(path).convert("L")
        except Exception:
            continue

        # 公告关闭（黑底浅色）：优先用二值模板；尺度/旋转收紧，避免卡顿
        if "announce_close" in base:
            rois = announce_rois(portrait)
            scales = (0.75, 1.0, 1.35, 1.8)
            # 已是旋转文件则不再现场旋转；原图才补四向
            if any(x in base for x in ("_90", "_180", "_270")):
                jobs.extend(_scaled_jobs(im, path, rois, scales))
            else:
                for vim, tag in (
                    (im, ""),
                    (im.transpose(Image.ROTATE_90), "#r90"),
                    (im.transpose(Image.ROTATE_180), "#r180"),
                    (im.transpose(Image.ROTATE_270), "#r270"),
                ):
                    jobs.extend(_scaled_jobs(vim, path, rois, scales, tag))
            continue

        # 刷么鱼菜单/玩法图标：多尺度 + 四向（竖屏窗）
        if base.startswith("moyu_"):
            if "menu" in base:
                rois = announce_rois(portrait)
            else:
                rois = continue_rois(portrait) if not portrait else [
                    (0.05, 0.15, 0.55, 0.70),
                    (0.35, 0.20, 0.55, 0.65),
                    (0.10, 0.50, 0.70, 0.45),
                ]
            scales = (0.55, 0.75, 1.0, 1.3, 1.7)
            jobs.extend(_scaled_jobs(im, path, rois, scales))
            if portrait or "menu" in base:
                for vim, tag in (
                    (im.transpose(Image.ROTATE_90), "#r90"),
                    (im.transpose(Image.ROTATE_270), "#r270"),
                ):
                    jobs.extend(_scaled_jobs(vim, path, rois, scales, tag))
            continue

        # 「继续游戏」按钮/竖排字模板
        if "continue_game" in base:
            if portrait and "land" in base and "port" not in base:
                continue
            if (not portrait) and "port" in base:
                continue
            rois = continue_rois(portrait)
            scales = (0.45, 0.6, 0.8, 1.0, 1.2)
            jobs.extend(_scaled_jobs(im, path, rois, scales))
            if portrait and "port" in base:
                jobs.extend(_scaled_jobs(
                    im.transpose(Image.ROTATE_90), path, rois, scales, "#r90"))
                jobs.extend(_scaled_jobs(
                    im.transpose(Image.ROTATE_270), path, rois, scales, "#r270"))
            continue

        if portrait:
            if "land" in base and "port" not in base:
                continue
            rois = port_rois()
            if "port" in base or "alt" in base or base in ("back.png", "back_user.png", "back_live.png"):
                jobs.append((im, 1.0, rois, path))
            elif "popup_close" in base or base.startswith("back_popup"):
                jobs.extend(_scaled_jobs(im, path, rois, (0.7, 0.85, 1.0, 1.2, 1.5, 2.0)))
                jobs.extend(_scaled_jobs(
                    im.transpose(Image.ROTATE_90), path, rois, (0.85, 1.0, 1.3), "#r90"))
                jobs.extend(_scaled_jobs(
                    im.transpose(Image.ROTATE_270), path, rois, (0.85, 1.0, 1.3), "#r270"))
        else:
            if "port" in base:
                continue
            rois = land_rois()
            if "popup_close" in base or base.startswith("back_popup"):
                jobs.extend(_scaled_jobs(im, path, rois, (0.7, 0.85, 1.0, 1.2, 1.5, 2.0)))
            else:
                jobs.append((im, 1.0, rois, path))
    if jobs:
        return jobs
    # 兜底：只有一张图时，竖屏才旋转一次
    for path in paths:
        if not path or not os.path.isfile(path):
            continue
        try:
            im = Image.open(path).convert("L")
        except Exception:
            continue
        if portrait:
            jobs.append((im.transpose(Image.ROTATE_270), 1.0, port_rois(), path + "#rot270"))
            jobs.append((im.transpose(Image.ROTATE_90), 1.0, port_rois(), path + "#rot90"))
        else:
            jobs.append((im, 1.0, land_rois(), path))
        break
    return jobs


def main():
    args = sys.argv[1:]
    roi_arg = None
    templates = []
    if "--roi" in args:
        i = args.index("--roi")
        roi_arg = tuple(float(args[i + k]) for k in range(1, 5))
        args = args[:i] + args[i + 5:]
    while "--template" in args:
        i = args.index("--template")
        if i + 1 < len(args):
            templates.append(args[i + 1])
            args = args[:i] + args[i + 2:]
        else:
            break
    if len(args) < 2 and not templates:
        print("usage: match.py hay.png needle.png [minScore] [--template more.png ...]", file=sys.stderr)
        return 1
    hay_path = args[0]
    if not templates:
        templates = [args[1]]
        min_score = float(args[2]) if len(args) >= 3 else 0.85
    else:
        if len(args) >= 2 and not args[1].replace(".", "", 1).isdigit():
            templates.insert(0, args[1])
            min_score = float(args[2]) if len(args) >= 3 else 0.85
        else:
            min_score = float(args[1]) if len(args) >= 2 else 0.85

    hay = Image.open(hay_path).convert("L")
    portrait = hay.size[1] > hay.size[0] * 1.08
    jobs = collect_templates(templates, portrait)
    if roi_arg:
        jobs = [(im, scale, [roi_arg], name) for im, scale, _, name in jobs]

    best = None
    weak = -1.0
    weak_name = ""
    for im, scale, rois, name in jobs:
        prep = prepare(im)
        if not prep:
            continue
        step = 1 if max(prep[4], prep[5]) <= 40 else 2
        for roi in rois:
            hit = search(hay, prep, roi, step=step)
            if not hit:
                continue
            if hit[0] > weak:
                weak = hit[0]
                weak_name = name
            if hit[0] < min_score:
                continue
            fx = (hit[1] + prep[4] / 2.0) / hay.size[0]
            fy = (hit[2] + prep[5] / 2.0) / hay.size[1]
            cand = (hit[0], fx, fy, scale, name)
            if best is None or cand[0] > best[0]:
                best = cand
            if best[0] >= 0.90:
                print("%.4f\t%.4f\t%.3f\t%.2f\t%s" % (best[1], best[2], best[0], best[3], os.path.basename(best[4])))
                return 0
    if not best:
        print("none\t%.3f\t%s" % (weak, os.path.basename(weak_name) if weak_name else ""))
        return 0
    print("%.4f\t%.4f\t%.3f\t%.2f\t%s" % (best[1], best[2], best[0], best[3], os.path.basename(best[4])))
    return 0


if __name__ == "__main__":
    sys.exit(main())
